WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsMenuViewModel.swift

```diff
@@ -63,14 +76,15 @@ class InPersonPaymentsMenuViewModel: ObservableObject {
         }
     }
 
-    private static func createPaymentGatewaysResultsController(siteID: Int64?) -> ResultsController<StoragePaymentGateway>? {
+    private static func createPaymentGatewaysResultsController(siteID: Int64?,
```

Comment:
Most likely we can go through this in our next pairing: Would be the same here if we declare `ResultsController<StoragePaymentGateway>` as lazy, and then make `createPaymentGatewaysResultsController()` non-static?

My guess of some the reasons to use static/lazy here would be that fetching from Storage could be an expensive operation and we want these to be kept through the lifetime of the object? I thought that static are initialized prior to program execution rather than when is accessed for first time (lazy), so it would make sense in this case, but I seem to find conflicting information if that's actually the case in Swift as well 🤔

Reply:
@iamgabrielma Thanks for the insightful question...

Key thing here which I think you've misread and might have misled you a bit: this is a static function, not a static property. If it were a `static let`, your reasoning that they're very similar but have different access characteristics makes more sense.

Since it's a function, there's no variable to init here. We just call it in the view model's init.

It's pretty similar, but not the same, and I think I originally made it a `lazy var`... but I prefer this way for a few reasons.

1. There's no such thing as `lazy let` in Swift, you have to make it a `lazy var`. It would be nice if we could say that the lazy property could only be set once, on first access, but we can't. I don't know the reasoning, but I guess it can only be a `var` due to the strict correctness of Swift: it's nothing until it's accessed, so it's changing from nothing to something when you first access it, so that means it's a `var`.
2. We know we're going to get the PaymentGateway from storage immediately, so we might as well get the storage set up straight away. Core Data does lots of asynchronous work, let's minimise any waiting time by getting set up when we create the class (I've not tested whether there is any, this is just based on feeling.)
3. It's generally easier to reason about code when there's less possibility of a time delay. By making this a `let` and setting it to a concrete value in `init`, future-us won't see the `lazy var` and have to wonder about any potential race conditions, things having changed in `dependencies` since init was called (they can't now, but that doesn't mean they won't ever be able to) or even have to think "when is this actually first called and created?"

## The real reason
I don't think I thought all that through at the time of writing the code. I think I just wanted it to be a let, not a var.

That means the reason I made the function static was so that I could use it before all the properties were set in the `init` call, since it's used to set a property, which must be done before you call any instance methods. It relies on stuff from `dependencies` to do that.

## Results Controllers are kept around so that we can observe changes
> My guess of some the reasons to use static/lazy here would be that fetching from Storage could be an expensive operation and we want these to be kept through the lifetime of the object?

I don't think results controllers are particularly expensive... I don't know that for sure though. Either way, they're our only option for what we use them for. 

We keep the same one around for a totally different reason: it's what allows us to be notified of changes in the data storage.

Note [where we use the `paymentGatewaysFetchedResultsController`, we set `onDidChangeContent` and `onDidResetContent` closures](https://github.com/woocommerce/woocommerce-ios/blob/5ae6721bb31c2e18eecbb537bb8216ec5ea7d350/WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person%20Payments/InPersonPaymentsMenuViewModel.swift#L55) which update the visual state of the toggle. That'll mean that whenever the underlying stored data for PaymentGateway for this site changes, we take a look for the `cod` gateway, and then update our toggle to match it.


Happy to go through further in pairing next week, or give me a shout this morning if you prefer 😊